script_runner/tasks.py

```diff
@@ -331,27 +340,60 @@ def _prepare_ssl_cert(ssl_cert_content):
 
 class OutputConsumer(object):
 
-    def __init__(self, out):
+    def __init__(self, ctx, out, name):
         self.out = out
-        self.buffer = StringIO()
+        self.name = name
+        self.ctx = ctx
+
+        logs_dir = self._get_log_dir()
+        try:
+            os.makedirs(logs_dir)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+
+        self.out_file = os.path.join(logs_dir, "{0}.{1}.{2}".format(
+            self.ctx.task_id, self.ctx.operation.retry_number, self.name))
+
         self.consumer = threading.Thread(target=self.consume_output)
         self.consumer.daemon = True
         self.consumer.start()
 
     def consume_output(self):
-        for line in iter(self.out.readline, b''):
-            self.buffer.write(line)
+        with open(self.out_file, 'w') as output:
+            for line in iter(self.out.readline, b''):
+                output.write(line)
+
         self.out.close()
 
     def join(self):
         self.consumer.join()
 
+    @staticmethod
+    def _get_log_dir():
+        return os.environ.get(CELERY_WORK_DIR_KEY, DEFAULT_TASK_LOG_DIR)
 
-class ProcessException(Exception):
 
-    def __init__(self, command, exit_code, stdout, stderr):
-        super(ProcessException, self).__init__(stderr)
+class ProcessException(Exception):
+    def __init__(self, command, exit_code, consumers):
         self.command = command
         self.exit_code = exit_code
-        self.stdout = stdout
-        self.stderr = stderr
+        self.consumers = consumers
+
+        message = 'command: {0}, exit_code: {1}, {2}'.format(
+            command, exit_code, ', '.join(['{0}: {1}'.format(
+                x, y.out_file) for x, y in consumers.iteritems()]))
+
+        super(ProcessException, self).__init__(message)
+
+    @property
+    def stdout(self):
+        return self._get_consumer_file('stdout')
+
+    @property
+    def stderr(self):
+        return self._get_consumer_file('stderr')
+
+    def _get_consumer_file(self, name):
+        with open(self.consumers[name].out_file, 'r') as f:
+            return StringIO(f.read()).getvalue()
```

Comment:
why additionally go through StringIO instead of `return f.read()`?

Reply:
Already responded to @jcollado: to comply with our existing unit tests. Maybe I should fix the unit tests instead?